File: backend/src/main/java/com/skewwhiffy/auraltester/notation/factory/InternalNotationFactory.java
package com.skewwhiffy.auraltester.notation.factory;

import com.skewwhiffy.auraltester.exception.UnrecognizedClefException;
import com.skewwhiffy.auraltester.notation.model.interval.DirectedInterval;
import com.skewwhiffy.auraltester.notation.model.key.Key;
import com.skewwhiffy.auraltester.notation.model.note.AbsoluteNote;
import lombok.AllArgsConstructor;
import org.springframework.stereotype.Service;
import com.skewwhiffy.auraltester.notation.model.clef.Clef;

import java.util.Arrays;
import java.util.List;
import java.util.Locale;

@Service
@AllArgsConstructor
public class InternalNotationFactory {
    private final ClefFactory clefFactory;
    private final NoteFactory noteFactory;
    private final IntervalFactory intervalFactory;
    private final KeyFactory keyFactory;

    public Clef clef(String clefRaw) {
        return switch (clefRaw.toLowerCase(Locale.UK)) {
            case "treble" -> clefFactory.getTreble();
            case "alto" -> clefFactory.getAlto();
            case "tenor" -> clefFactory.getTenor();
            case "bass" -> clefFactory.getBass();
            default -> throw new UnrecognizedClefException(clefRaw);
        };
    }

    public AbsoluteNote getNote(String noteRaw) {
        return noteFactory.getAbsoluteNote(noteRaw);
    }

    public List<AbsoluteNote> getNotes(String notesRaw) {
        return Arrays.stream(notesRaw.split(" "))
                .map(this::getNote)
                .toList();
    }

    public DirectedInterval getDirectedInterval(String rawInterval) {
        return intervalFactory.getDirectedInterval(rawInterval);
    }

    public List<DirectedInterval> getDirectedIntervals(String rawIntervals) {
        return intervalFactory.getDirectedIntervals(rawIntervals);
    }

    public Key getKey(String rawKey) {
        return keyFactory.getKey(rawKey);
    }
}

File: backend/src/main/kotlin/com/skewwhiffy/auraltester/notation/factory/InternalNotationFactory.kt
package com.skewwhiffy.auraltester.notation.factory

import com.skewwhiffy.auraltester.exception.UnrecognizedClefException
import com.skewwhiffy.auraltester.notation.model.clef.Clef
import com.skewwhiffy.auraltester.notation.model.interval.DirectedInterval
import org.springframework.stereotype.Service

@Service
class InternalNotationFactory(
    private val clefFactory: ClefFactory,
    private val noteFactory: NoteFactory,
    private val intervalFactory: IntervalFactory,
    private val keyFactory: KeyFactory
) {
    fun clef(clefRaw: String): Clef {
        return when (clefRaw.lowercase()) {
            "treble" -> clefFactory.treble
            "alto" -> clefFactory.alto
            "tenor" -> clefFactory.tenor
            "bass" -> clefFactory.bass
            else -> throw UnrecognizedClefException(clefRaw)
        }
    }

    fun getNote(noteRaw: String) = noteFactory.getAbsoluteNote(noteRaw)

    fun getNotes(notesRaw: String) = notesRaw.split(" ").map(::getNote)

    fun getDirectedInterval(rawInterval: String) = intervalFactory.getDirectedInterval(rawInterval)

    fun getDirectedIntervals(rawIntervals: String) = intervalFactory.getDirectedIntervals(rawIntervals)

    fun getKey(rawKey: String) = keyFactory.getKey(rawKey)
}
